### ISSUE TITLE
perf: enable mx.compile for training by default — 16% speedup on Qwen3-0.6B

### ISSUE THREAD

**Issue by teilomillet** [teilomillet/textpolicy#40]
## Problem

`mx.compile` provides a **16% speedup** on Qwen3-0.6B training steps but is disabled by default (`compile_training=False`). It was disabled because it causes a 4x slowdown on TinyLM (50K params), but this is irrelevant for real models.

### Evidence

```
Qwen3-0.6B (596M params), 4 episodes × 128 tokens:
  Uncompiled: 1,069 ms/step
  Compiled:     918 ms/step  → 1.16x speedup

TinyLM (50K params):
  Uncompiled:   5.8 ms/step
  Compiled:    25.1 ms/step  → 0.23x (4x SLOWER)
```

`mx.compile` traces the computation graph and optimizes it. For large models, the compilation overhead is amortized over the many operations. For tiny models, the overhead exceeds the benefit.

### Root cause

`textpolicy/training/trainer.py` constructor defaults to `compile_training=False`. When `True`, it wraps `_loss_fn` with `mx.compile(nn.value_and_grad(...))`.

### 1. Auto-detect whether to compile
Based on model parameter count — compile when params > threshold (e.g., 1M).

### 2. Verify correctness under compilation
`mx.compile` has constraints: no Python conditionals on `mx.array` values inside traced functions, no `mx.eval()` calls. The `_loss_fn` must respect these.

### 3. Verify no silent numerical drift
Compiled and uncompiled paths must produce the same loss values.

### L1: Verify _loss_fn is compile-safe
```python
@pytest.mark.requires_model
def test_loss_fn_is_compile_safe():
    """_loss_fn produces identical loss compiled vs uncompiled."""
    model, tokenizer = load_model('Qwen/Qwen3-0.6B')
    batch = build_real_batch(model, tokenizer, n_episodes=4, max_tokens=32)
    
    # Uncompiled
    trainer_uc = Trainer(model=model, ..., compile_training=False)
    m_uc = trainer_uc.train(batch)
    
    # Compiled
    trainer_c = Trainer(model=model, ..., compile_training=True)
    m_c = trainer_c.train(batch)
    
    # Loss must match within tolerance
    assert abs(m_uc['loss'] - m_c['loss']) < 0.01
```

### L2: Verify no Python-conditional-on-mx.array violations
```bash
# Search for patterns that break under mx.compile:
grep -n 'if.*\.item()' textpolicy/training/trainer.py
grep -n 'if.*mx\.' textpolicy/algorithms/grpo.py
```

**Expected**: No `.item()` calls or Python `if` on `mx.array` values inside `_loss_fn` or any function it calls. If found, refactor to use `mx.where`.

### L3: Compile speedup scales with model size
```python
@pytest.mark.unit
def test_compile_speedup_scales_with_model_size():
    """Compilation overhead is amortized for models above threshold."""
    for dim, expected_faster in [(32, False), (256, True), (512, True)]:
        model = TinyLM(dim=dim)
        # ... measure compiled vs uncompiled ...
        # Small models: compiled should be slower or equal
        # Large models: compiled should be faster
```

### L4: Auto-detection threshold
```python
@pytest.mark.unit
def test_auto_compile_threshold():
    """Trainer auto-enables compile for models above parameter threshold."""
    small = TinyLM(dim=32)   # ~5K params → should NOT compile
    large = TinyLM(dim=512)  # ~530K params → should compile
    
    trainer_s = Trainer(model=small, ..., compile_training='auto')
    trainer_l = Trainer(model=large, ..., compile_training='auto')
    
    assert trainer_s._compiled is False
    assert trainer_l._compiled is True
```

### L5: Compile + grad_clip interaction
```python
@pytest.mark.requires_model
def test_compile_with_grad_clip():
    """mx.compile doesn't interfere with gradient clipping."""
    # Run 10 steps compiled with clip, check gradients are finite
    # and loss decreases
```

## Acceptance criteria

- [ ] L1 confirms numerical parity between compiled and uncompiled
- [ ] L2 confirms no compile-unsafe patterns in the hot path
- [ ] L3 identifies the crossover point where compile helps
- [ ] L4 implements and tests auto-detection
- [ ] L5 verifies compile + clip interaction
- [ ] Default changed to `compile_training='auto'`
- [ ] No regression in test suite
